Add extra-languages config key

mazen-asef opened the pull request:
app: Add new default-languages key

In order to have user-settable languages in the configuration files
so users can add desired locales pertinent to their interests, without
disabling the automatic AccountServices dbus calls that happens when you
have xa.languages set in the configuration files, we created the
default-languages key.

This is inline with #https://github.com/flatpak/flatpak/issues/3043 and
https://gitlab.gnome.org/GNOME/gnome-software/issues/697.

mazen-asef commented:
CC @mwleeds @pwithnall @ramcq

ramcq commented on common/flatpak-dir.c:
```diff
@@ -14082,17 +14082,24 @@ char **
   if (config)
     {
-      char **langs = g_key_file_get_string_list (config, "core", "xa.languages", NULL, NULL);
+      printf("itshere");
```
Debug print can be removed... :P

mazen-asef replied:
RIght, sorry hahah

mazen-asef commented:
Ah, seems reasonable - I was under the impression that we wouldn't actually use them both at the same time. Will make it so it actually get the info from all sides.

wjt commented on common/flatpak-dir.c:
```diff
@@ -14082,10 +14082,15 @@ char **
   if (config)
     {
-      char **langs = g_key_file_get_string_list (config, "core", "xa.languages", NULL, NULL);
+      if (g_key_file_has_key (config, "core", "xa.default-languages", NULL))
```
How does setting `xa.default-languages` differ from setting `xa.languages`?

ramcq commented:
> Ah, seems reasonable - I was under the impression that we wouldn't actually use them both at the same time. Will make it so it actually get the info from all sides.

It doesn't make sense for them to be used at the same time, no, but this alone isn't a reason to have two keys with the same behaviour. The reason we're adding a new key is because we want it to work differently...

mazen-asef commented:
Now the `flatpak_dir_get_locale_languages` merges and de-dups whatever is stored in both xa.langauges and xa.default-languages, without giving any precedence to each of them, but both are still above AccountServices calls.

wjt commented:
> Now the `flatpak_dir_get_locale_languages` merges and de-dups whatever is stored in both xa.langauges and xa.default-languages, without giving any precedence to each of them, but both are still above AccountServices calls.

But that's still not achieving the goal of having different behaviour to `xa.languages`. The behaviour we want (AIUI) is:

* If `xa.languages` is set, use these, and no others
* Otherwise, take the union of:
  * `xa.default-languages`, if set
  * `flatpak_dir_get_default_locale_languages()`: for system repo, the system default locales; for user repo, the langs based on the user's locale

mazen-asef commented:
Sorry @wjt, just after the PR i saw the mistake.

I couldn't make the join's neater though, it look kinda messy having to cast `char **` to make the merge of the two arrays for the `GPtrArray`.

Is there a cleaner way to do it?

mazen-asef commented:
Also, I thought we could be using `xa.languages` and `xa.default-languages` if available, isn't it?

mazen-asef commented:
After discussion with @ramcq it's better if we just keep the current semantic for `xa.languages` and if that's not set, use the `xa.default-languages` instead, on top of the other standard methods for fetching languages on `flatpak_dir_get_default_locale_languages()`

mwleeds commented on common/flatpak-installation.c:
```diff
@@ -1648,6 +1648,33 @@ flatpak_installation_get_config (FlatpakInstallation *self,
+ * @error: return location for a #GError
+ *
+ * Get the default languages being used by the user/system
+ * on the ostree configuration files, stored either by xa.languages
```
This doesn't quite match the implementation. You could say "For user installations, this is determined from the `xa.languages` config key, or if unset the combination of `xa.default-languages` and languages set by environment variables. For system installations, if `xa.languages` is unset, the combination of `xa.default-languages` and the list returned by `AccountsService` is used."

Or just don't specify how it's implemented.

mwleeds commented on common/flatpak-dir.c:
```diff
@@ -14075,20 +14096,17 @@ flatpak_dir_get_default_locale_languages (FlatpakDir *self)
-        return sort_strv (langs);
-    }
+  langs = flatpak_dir_get_language_config_key (self, "xa.languages");
+  if (langs)
```
Maybe add a comment that an empty array is intended to mean "all languages", and direct people to the `flatpak-config` man page?

mwleeds commented on common/flatpak-installation.c:
```diff
@@ -1648,6 +1648,33 @@ flatpak_installation_get_config (FlatpakInstallation *self,
+ * and xa.default-languages keys, or from AccountServices dbus.
+ *
+ * Returns: (transfer none): Default languages array of strings, or an
+ * 0 initializied char * struct.
```
And the second line should be indented two spaces

pwithnall commented on common/flatpak-installation.c:
```diff
@@ -1648,6 +1648,32 @@ flatpak_installation_get_config (FlatpakInstallation *self,
+ * to display.
+ *
+ * Returns: (array zero-terminated=1) (element-type utf8) (transfer full):
+ *   A possibly empty array of locale strings, or %NULL on error.
```
Needs a `Since: x` line declaring which version of libflatpak the API first appeared in.

mwleeds commented on common/flatpak-installation.c:
```diff
@@ -1648,6 +1648,33 @@ flatpak_installation_get_config (FlatpakInstallation *self,
+ *
+ * Returns: (array zero-terminated=1) (element-type utf8) (transfer full):
+ *   A possibly empty array of locale strings, or %NULL on error.
+ * Since: 1.4
```
Should be `Since: 1.4.3` I think. That will be the first release to have this API.

matthiasclasen commented on doc/flatpak-config.xml:
```diff
@@ -62,7 +62,17 @@
+                <listitem><para>
+                   This key is used when languages is not set, and it defines locale
+                   extensions on top of the system configured languages. The value is a
+                   semicolon-separated list of two-letter  language codes, or one of the
```
There's an extra space here between 'two-letter' and 'language'

mwleeds commented on common/flatpak-dir.c:
```diff
@@ -14047,17 +14047,38 @@ sort_strv (char **strv)
 }
 
+static char **
+flatpak_dir_get_config_strv (FlatpakDir *self, char *key) {
```
The curly bracket should be on the next line

mwleeds commented:
@rh-atomic-bot r+

mwleeds commented:
Looks like the newly added unit test failed. Mazen is looking into it.

alexlarsson commented:
Argh. Seems I commented in the wrong commit. I'll add it here.

I don't think the name "default-languages" for the new config key is right. The default is still the system-wide default taken from the account service. This is more in *addition* to that, so I think a better name would be extra-languages. Also, since that is additive it shouldn't really support the magic support for `*` to mean all languages, etc. If you want that you need to use `languages`.

More review details in the commit

alexlarsson commented on app/flatpak-builtins-config.c:
```diff
@@ -112,6 +112,7 @@ typedef struct
 
 ConfigKey keys[] = {
   { "languages", parse_lang, print_lang, get_lang_default },
+  { "default-languages", parse_lang, print_lang, get_lang_default },
```
This isn't right, we should not handle the special `*` or `all` things for this setting, and there should not be a need to print a default value for it. The default of this particular key is "unset", and if you put anything in it that will normally be visible in the default output of the `languages` key. This also seems to imply that we should list the extra-languages key before the languages key.

mazen-asef replied:
I see - yeah, it makes sense. However, i didn't understand how this specific piece isn't right - if I want extra-languages to be set through the cli this line should stay here, doesn't it?

mwleeds replied:
Not in its current form, because the implementations of `parse_lang()` and `print_lang()` interpret the `*`/`all` values. And instead of `get_lang_default()` for extra-languages the default is unset.

alexlarsson commented on common/flatpak-dir.c:
```diff
@@ -14068,27 +14089,24 @@ flatpak_dir_get_default_locale_languages (FlatpakDir *self)
 
-  /* If none were found, fall back to using all languages */
-  if (langs->len == 0)
-    return g_new0 (char *, 1);
```
Why are you deleting this? If we have no account service, and if extra-languages is not set, i don't think using only the current locale is the right thing. That may be different to the thing someone eventually uses. For example locale might be C or en_US during a system install, but then change to something else.

mwleeds replied:
If `extra-languages` is unset and no languages were found through localed or AccountsService, we are still using all languages. In that case the `flatpak_strv_merge` below will return an empty array.

mazen-asef replied:
You're right, will remove this change.

mwleeds replied:
No I was explaining why I think we should should leave this `if (langs->len == 0) return g_new0 (char *, 1);` deleted.

alexlarsson commented on common/flatpak-installation.c:
```diff
@@ -1648,6 +1648,33 @@ flatpak_installation_get_config (FlatpakInstallation *self,
+ * Returns: (array zero-terminated=1) (element-type utf8) (transfer full):
+ *   A possibly empty array of locale strings, or %NULL on error.
+ * Since: 1.4.3
+ */
```
Should be since 1.5.0, we're not merging this to the stable branch

alexlarsson commented on common/flatpak-installation.c:
```diff
@@ -1648,6 +1648,33 @@ flatpak_installation_get_config (FlatpakInstallation *self,
+ * subpaths to install of locale extensions. This list may also be used
+ * by frontends like GNOME Software to decide which language-specific apps
+ * to display.
+ *
```
This should mention that an empty array means all languages are installed.

alexlarsson commented on doc/flatpak-config.xml:
```diff
@@ -62,7 +62,17 @@
+                   extensions on top of the system configured languages. The value is a
+                   semicolon-separated list of two-letter language codes, or one of the
+                   special values * or all. If this key is unset, this defaults to system
+                   configured languages.
```
We should not support (or document) the `*`/`all` feature for this, it makes no sense for something that is additive, nor will it really work (as all means making this empty, which will merge with the system one causing it to not be empty anymore). If you need that override the full `languages` keys.

mazen-asef commented:
@alexlarsson @mwleeds I have updated the PR with your siggestions - also fixed the test issues and found a new tweak to be made on the `flatpak-system-helper.c`

let me know what do you think :)

alexlarsson commented on common/flatpak-dir.c:
```diff
@@ -14068,27 +14088,24 @@ flatpak_dir_get_default_locale_languages (FlatpakDir *self)
   g_ptr_array_add (langs, NULL);
 
-  return (char **) g_ptr_array_free (g_steal_pointer (&langs), FALSE);
+  return flatpak_strv_merge (extra_languages, (char **) langs->pdata);
```
All the other codepaths sort the result, but this merge doesn't.

mazen-asef replied:
Oh yeah! Sorry - I have deleted the previous sort for langs and added `sort_strv` for the return on `flatpak_strv_merge`

mwleeds commented on common/flatpak-dir.c:
```diff
@@ -14047,17 +14047,37 @@ sort_strv (char **strv)
 
   if (flatpak_dir_is_user (self))
-    return flatpak_get_current_locale_langs ();
+    return flatpak_strv_merge (extra_languages, flatpak_get_current_locale_langs ());
```
We need a `sort_strv` here too right?

alexlarsson replied:
Yeah.

mwleeds commented on app/flatpak-builtins-config.c:
```diff
@@ -86,12 +82,28 @@ parse_lang (const char *value, GError **error)
+      strcmp (value, "*all*") == 0)
+    return g_strdup ("");
+
+  return parse_lang(value, error);
```
missing space before left parentheses

mazen-asef commented:
Updated with the latest suggestions :)

alexlarsson commented:
@rh-atomic-bot r+

alexlarsson commented:
Some test is failing:
```
PASS: testlibrary 6 /library/installation-config
ERROR: testlibrary - Bail out! flatpak:ERROR:tests/testlibrary.c:275:test_languages_config: assertion failed (value[0] == "en"): (NULL == "en")
```
Can you take a look?

mwleeds commented:
> Ah, interesting - I made the assumption the local languages used on the testing suite was 'en' since there were no further data, (mine were all green), but it seems the test doesn't have anything said. That being so, should I just remove this test, or should I compare it to NULL?

I would recommend doing `g_setenv ("LANGUAGE", "en", TRUE);` in `global_setup()` where all the other environment variables are set. That way `g_get_language_names()` will pick up English and you can leave the test as written.

Ideally we'd test the behavior of system installations too when `xa.languages` is unset, but that would require creating mock `localed` and `Accounts` D-Bus services so probably not worth the effort.

mwleeds commented:
@rh-atomic-bot r+
